### PR TITLE
Filtering improvements

### PR DESCRIPTION
**Pull request by mario**
This improves filtering:

- add rotation support
- improves performance in case of many files (results limited to 100)
- no more duplicates
- add a few usability improvements (like hiding FAB & footer when searching)

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
If you enter a query which leads to no results, the message is confusing.
The text should be something like "no results for $query in this folder".
screenshot_1486201042

**Comment by mario** [comment]
@AndyScherzinger can you get us a nice light blue search icon please? :)

**Comment by AndyScherzinger** [comment]
sure!
Color? -> system bar, action bar or the light blue of the folder in the empty content view, as seen above?

**Comment by mario** [comment]
The folder in the empty content view blue :)

**Comment by AndyScherzinger** [comment]
rebased... just pull :) --> https://github.com/nextcloud/android/pull/609/commits/f66601fdf34bbca138b2d43d9335b33b645b5644

**Comment by AndyScherzinger** [comment]
@mario can you post a screenshot with the UI for @jancborchardt and/or @eppfel to review when feature complete? :)

**Comment by mario** [comment]
@AndyScherzinger yes.

**Comment by AndyScherzinger** [comment]
Hope the icon is fine, I've chosen the standard menu icon size (not sure where you want to use it)

**Comment by mario** [comment]
@AndyScherzinger I need it the same size as that folder in the image above :)

**Comment by AndyScherzinger** [comment]
Alright, will have the change pushed in a couple of minutes :)

**Comment by AndyScherzinger** [comment]
There you go :) https://github.com/nextcloud/android/pull/609/commits/dce0bf96f07837363fe9a013a422a69fc20748b1

File: res/values/strings.xml [hunk 1]
```diff
@@ -79,7 +79,9 @@
     <string name="uploader_upload_files_behaviour_upload_and_delete_from_source">Delete file from source folder</string>
     <string name="file_list_seconds_ago">seconds ago</string>
     <string name="file_list_empty_headline">No files in here</string>
+    <string name="file_list_empty_headline_search">No results in this folder</string>
     <string name="file_list_empty">Upload some content or sync with your devices!</string>
+    <string name="file_list_empty_search">Maybe try looking in another one? :)</string>
```
Comment (AndyScherzinger) [inline res/values/strings.xml]: Please be serious :P

**Comment by mario** [comment]
Say what? :p

**Comment by AndyScherzinger** [comment]
Well, where is this text shown? Can "one" be replaced by a more explicit term (do you mean folder?)

**Comment by mario** [comment]
Yup, a folder.

It is shown when there are no results.

**Comment by mario** [comment]
screenshot_20170207-165030

@jancborchardt @eppfel  :)

@AndyScherzinger this branch is now ready.

File: res/values/strings.xml [hunk 2]
```diff
@@ -81,7 +81,7 @@
     <string name="file_list_empty_headline">No files in here</string>
     <string name="file_list_empty_headline_search">No results in this folder</string>
     <string name="file_list_empty">Upload some content or sync with your devices!</string>
-    <string name="file_list_empty_search">Maybe try looking in another one? :)</string>
+    <string name="file_list_empty_search">Try looking in another folder? :)</string>
```
Comment (AndyScherzinger) [inline res/values/strings.xml]: can we get rid of the smiley? :)

Comment (mario) [inline res/values/strings.xml]: No :P Where's the fun in removing all the little joys in life? :)

Comment (eppfel) [inline res/values/strings.xml]: Smilies are not funny, sry. ;)

**Comment by eppfel** [comment]
Just for understanding do i get results only for the folder i searched in?

File: res/values/strings.xml (hunk 2 above)
Comment (AndyScherzinger) [inline res/values/strings.xml]: Well, this is an app in the Nextcloud eco systems, so you do have paying, enterprise customers!

**Comment by mario** [comment]
@eppfel yes, this PR is not about introducing real search, but for fixing the filter which is broken.

File: res/values/strings.xml (hunk 2 above)
Comment (mario) [inline res/values/strings.xml]: And enterprises don't deserve fun? :-/ Removed.

**Comment by AndyScherzinger** [comment]
Since I am not a regular user of this feature, what was broken? :)

**Comment by mario** [comment]
a) Rotation didn't preserve search nor the search query 
b) When filtering, I would get countless results (if 3 items were results, I'd get 3 of them infinite times)
c) Also some other minor UX stuff (i.e. FAB handling)

File: res/values/strings.xml [hunk 3]
```diff
@@ -81,7 +81,7 @@
     <string name="file_list_empty_headline">No files in here</string>
     <string name="file_list_empty_headline_search">No results in this folder</string>
     <string name="file_list_empty">Upload some content or sync with your devices!</string>
-    <string name="file_list_empty_search">Try looking in another folder? :)</string>
+    <string name="file_list_empty_search">Try looking in another folder?</string>
```
Comment (eppfel) [inline res/values/strings.xml]: :)

Comment (mario) [inline res/values/strings.xml]: I made a change you wanted, now approve :P

**Comment by mario** [comment]
@AndyScherzinger @tobiasKaminsky ready for merge?

**Comment by AndyScherzinger** [comment]
👍 @mario code wise fine with me, I did a slight change though, see https://github.com/nextcloud/android/pull/609/commits/053b2d126c7484fa9e3267ecb821b644eba647eb

before re-adding ```|collapseActionView``` (overflow menu gets slightly pushed to the right (completely hidden on your screenshot, thus device resolution dependent...)
device-2017-02-07-234101

after my commit(look&feel like before):
device-2017-02-07-235306

**Comment by AndyScherzinger** [comment]
@mario are you also fine with my latest commit?
@tobiasKaminsky can you review/approve?

**Comment by mario** [comment]
👍  @AndyScherzinger all good.

**Comment by tobiasKaminsky** [comment]
Rotation resets search query in header, but preserves filtered items.
In the second screenshot the query "06" is missing, but the filtered items are correct. 
screenshot_1486739882
screenshot_1486739893

**Comment by tobiasKaminsky** [comment]
FAB never comes back after search
- search something
- press back to close search, or/and leave folder
- FAB is gone forever

**Comment by tobiasKaminsky** [comment]
View change shows shortly "no files found"
- show something that leads to only one item
- change from list to grid view
- for a very short time the "no files found" screen is shown

**Comment by tobiasKaminsky** [comment]
All was tested on an emulator with Nougat.

**Comment by tobiasKaminsky** [comment]
:-1: until the bugs are fixed

**Comment by mario** [comment]
"View change shows shortly "no files found"

This is a problem all over the app, not just this part and needs more changes than this PR warrants.

I'll look into the other bugs.

**Comment by mario** [comment]
@AndyScherzinger you broke search query persistance support xD Fixing it now, or at least trying :D

**Comment by mario** [comment]
Fixed all.

**Comment by tobiasKaminsky** [comment]
I think the footer should be visible during filtering so that the user can see how many filtered items exists, e.g. to find out how many images are from 201701.

**Comment by tobiasKaminsky** [comment]
BUG:
Overflow menu is not visible during filtering. This worked already in https://github.com/nextcloud/android/pull/609#issuecomment-278972231

**Comment by mario** [comment]
@tobiasKaminsky  I will happily restore the footer, but how is the lack of overflow menu a bug? Why do you want to click the overflow menu while filtering?

If you do a select (long click on one of the items), the menu is properly shown. Please try :)

**Comment by tobiasKaminsky** [comment]
BUG:
- filter for exactly one char that does not exist, e.g. "/"
- nice transition of "no results" 👍 
- remove one char
- transition overlaps with file list

**Comment by tobiasKaminsky** [comment]
> Why do you want to click the overflow menu while filtering? 

to change from file list to grid view, e.g. you are filtering for certain images and want to have them in grid view. Without overflow you would have to remove the query, change to grid, filter again.

> If you do a select (long click on one of the items), the menu is properly shown. 

In the overflow menu of a folder view is "refresh account", "sort", "list view". These are not the menu accessible by long click.

**Comment by mario** [comment]
@tobiasKaminsky if you're filtering, you probably should already be in the right view and not switch while filtering.

**Comment by jancborchardt** [comment]
Looks good, but the search icon should be grey instead of blue. :)

(Also I guess in the future the results from other folders will be loaded automatically, like on the server. ;)

**Comment by mario** [comment]
@jancborchardt all other icons in the list are blue - why not this one? When it's grey, it blends in too much...

**Comment by jancborchardt** [comment]
Which list? I was talking about the "No results in this folder" message. Just like on the server, the search icon should be grey there. :)

**Comment by AndyScherzinger** [comment]
@jancborchardt basically this one for example https://github.com/nextcloud/android/pull/609#issuecomment-277431876 while ignoring the search part. The light blue folder is also used in other parts of the app, eg for displaying an empty remote folder,like it is also done on the Web UI.

**Comment by tobiasKaminsky** [comment]
Further discussion about the color please in #661

**Comment by AndyScherzinger** [comment]
@mario can the branch be deleted?

**Comment by AndyScherzinger** [comment]
@mario can the branch be deleted?

**Comment by mario** [comment]
Yes

**Pull request closed without merging**